File: jsol-compiler-py/0400-compiler-utils.py
import math
import functools
from jsol_core import JSOL

# @JSOL v0.2.97 - AST-Free Global Utilities (1/2)

def bComp_IsWordChar(saCh): 

  if saCh == "": 

    return False;


  iCode = ord(saCh[ 0]);
  if iCode >= 48 and iCode <= 57: 

    return True;


  if iCode >= 65 and iCode <= 90: 

    return True;


  if iCode >= 97 and iCode <= 122: 

    return True;


  if iCode == 95: 

    return True;


  return False;


def mComp_ParseArgs(saCode, iOpenParen): 

  iParenCount = 1;
  iBracketCount = 0;
  iBraceCount = 0;
  bInStr = False;
  iCloseParen = -1;
  aArgs = [];
  iCurrentArgStart = iOpenParen + 1;
  iRLen = len(saCode);

  i = iOpenParen + 1;
  while i < iRLen: 

    saChar = saCode[( i):( i)+( 1)];
    saPrev = saCode[( i - 1):( i - 1)+( 1)];
    if saChar == "\"" and saPrev != "\\": 

      bInStr = not bInStr;


    if bInStr == False: 

      if saChar == "(": 

        iParenCount = iParenCount + 1;


      if saChar == ")": 

        iParenCount = iParenCount - 1;


      if saChar == "[": 

        iBracketCount = iBracketCount + 1;


      if saChar == "]": 

        iBracketCount = iBracketCount - 1;


      if saChar == "{": 

        iBraceCount = iBraceCount + 1;


      if saChar == "}": 

        iBraceCount = iBraceCount - 1;




    if saChar == "," and iParenCount == 1 and iBracketCount == 0 and iBraceCount == 0 and bInStr == False: 

      aArgs.append( saCode[( iCurrentArgStart):( iCurrentArgStart)+( i - iCurrentArgStart)]);
      iCurrentArgStart = i + 1;


    elif iParenCount == 0: 

      aArgs.append( saCode[( iCurrentArgStart):( iCurrentArgStart)+( i - iCurrentArgStart)]);
      iCloseParen = i;
      break;


    i = i + 1;


  return JSOL.dict("close",  iCloseParen,  "args",  aArgs);


def iComp_FindCloseBrace(saCode, iOpenBrace): 

  iBraceCount = 1;
  iRLen = len(saCode);
  i = iOpenBrace + 1;
  while i < iRLen: 

    saChar = saCode[( i):( i)+( 1)];
    if saChar == "{": 

      iBraceCount = iBraceCount + 1;


    if saChar == "}": 

      iBraceCount = iBraceCount - 1;


    if iBraceCount == 0: 

      return i;


    i = i + 1;


  return -1;


def iComp_FindCloseParen(saCode, iOpenParen): 

  iParenCount = 1;
  iRLen = len(saCode);
  i = iOpenParen + 1;
  while i < iRLen: 

    saChar = saCode[( i):( i)+( 1)];
    if saChar == "(": 

      iParenCount = iParenCount + 1;


    if saChar == ")": 

      iParenCount = iParenCount - 1;


    if iParenCount == 0: 

      return i;


    i = i + 1;


  return -1;


def iComp_FindStmtEnd(saCode, iStart): 

  iRLen = len(saCode);
  iEndIdx = iStart;
  bFindingEnd = True;
  while iEndIdx < iRLen and bFindingEnd == True: 

    saChar = saCode[( iEndIdx):( iEndIdx)+( 1)];
    if saChar == " " or saChar == "\n" or saChar == "\r" or saChar == ")" or saChar == ";": 

      iEndIdx = iEndIdx + 1;


    else: 

      bFindingEnd = False;




  return iEndIdx;

File: jsol-compiler-node.bak/0400-compiler-utils.js
// @JSOL v0.2.97 - AST-Free Global Utilities (1/2)

const $bComp_IsWordChar = function($saCh) {
  if ($saCh === "") {
    return false;
  }
  const $iCode = Str["char"]($saCh,  0);
    if ($iCode >= 48 && $iCode <= 57) {
    return true;
  }
  if ($iCode >= 65 && $iCode <= 90) {
    return true;
  }
  if ($iCode >= 97 && $iCode <= 122) {
    return true;
  }
  if ($iCode === 95) {
    return true;
  }
  return false;
};
const $mComp_ParseArgs = function($saCode, $iOpenParen) {
  let $iParenCount = 1;
    let $iBracketCount = 0;
    let $iBraceCount = 0;
    let $bInStr = false;
    let $iCloseParen = -1;
    let $aArgs = [];
    let $iCurrentArgStart = $iOpenParen + 1;
    const $iRLen = Str["len"]($saCode);

    for (let $i = $iOpenParen + 1; $i < $iRLen; $i = $i + 1) {
    const $saChar = Str["sub"]($saCode,  $i,  1);
        const $saPrev = Str["sub"]($saCode,  $i - 1,  1);
        if ($saChar === "\"" && $saPrev !== "\\") {
      $bInStr = !$bInStr;
    }
    if ($bInStr === false) {
      if ($saChar === "(") {
        $iParenCount = $iParenCount + 1;
      }
      if ($saChar === ")") {
        $iParenCount = $iParenCount - 1;
      }
      if ($saChar === "[") {
        $iBracketCount = $iBracketCount + 1;
      }
      if ($saChar === "]") {
        $iBracketCount = $iBracketCount - 1;
      }
      if ($saChar === "{") {
        $iBraceCount = $iBraceCount + 1;
      }
      if ($saChar === "}") {
        $iBraceCount = $iBraceCount - 1;
      }
    }
    if ($saChar === "," && $iParenCount === 1 && $iBracketCount === 0 && $iBraceCount === 0 && $bInStr === false) {
      $aArgs.push( Str["sub"]($saCode,  $iCurrentArgStart,  $i - $iCurrentArgStart));
            $iCurrentArgStart = $i + 1;
    }
    else if ($iParenCount === 0) {
      $aArgs.push( Str["sub"]($saCode,  $iCurrentArgStart,  $i - $iCurrentArgStart));
            $iCloseParen = $i;
            break;
    }
  }
  return JSOL.dict("close",  $iCloseParen,  "args",  $aArgs);
};
const $iComp_FindCloseBrace = function($saCode, $iOpenBrace) {
  let $iBraceCount = 1;
    const $iRLen = Str["len"]($saCode);
    for (let $i = $iOpenBrace + 1; $i < $iRLen; $i = $i + 1) {
    const $saChar = Str["sub"]($saCode,  $i,  1);
        if ($saChar === "{") {
      $iBraceCount = $iBraceCount + 1;
    }
    if ($saChar === "}") {
      $iBraceCount = $iBraceCount - 1;
    }
    if ($iBraceCount === 0) {
      return $i;
    }
  }
  return -1;
};
const $iComp_FindCloseParen = function($saCode, $iOpenParen) {
  let $iParenCount = 1;
    const $iRLen = Str["len"]($saCode);
    for (let $i = $iOpenParen + 1; $i < $iRLen; $i = $i + 1) {
    const $saChar = Str["sub"]($saCode,  $i,  1);
        if ($saChar === "(") {
      $iParenCount = $iParenCount + 1;
    }
    if ($saChar === ")") {
      $iParenCount = $iParenCount - 1;
    }
    if ($iParenCount === 0) {
      return $i;
    }
  }
  return -1;
};
const $iComp_FindStmtEnd = function($saCode, $iStart) {
  const $iRLen = Str["len"]($saCode);
    let $iEndIdx = $iStart;
    let $bFindingEnd = true;
    while ($iEndIdx < $iRLen && $bFindingEnd === true) {
    const $saChar = Str["sub"]($saCode,  $iEndIdx,  1);
        if ($saChar === " " || $saChar === "\n" || $saChar === "\r" || $saChar === ")" || $saChar === ";") {
      $iEndIdx = $iEndIdx + 1;
    }
    else {
      $bFindingEnd = false;
    }
  }
  return $iEndIdx;
};
